plugins/hls-haddock-comments-plugin/src/Ide/Plugin/HaddockComments/Data.hs

```diff
@@ -0,0 +1,168 @@
+            dpCol <- flip fmap getAnnsT $ \anns ->
+                case sameLineAsPrev of
+                    Just prevNode' ->
+                        -- If the previous node is the anchor, using 0 as column will make current code align with
+                        -- the previous one.
+                        -- Otherwise, use the column of entry delta of the previous node.
+                        -- The map lookup should not fail. '2' is used as a fallback value to make sure the syntax
+                        -- is correct after the changes.
+                        if usePrevNodeAsAnchor then 0 else maybe 2 (deltaColumn . annEntryDelta)
+                            $ anns Map.!? mkAnnKey prevNode'
+                    -- We subtract the real column to get dp column.
+                    Nothing -> (maybe 2 srcSpanStartCol . realSpan $ getLoc node) - anchorCol
+            -- Modify the current node
+            modifyAnnsT $
+                let updateCurrent :: Annotation -> Annotation
+                    updateCurrent ann = ann {
+                            -- If there exist non-haddock comments, we simply inherit the first one's delta pos,
+                            -- and move them two lines below, to seperate them from our newly added haddock comments
+                            -- Otherwise, inherit the node's entry delta pos.
+                            annPriorComments = case annPriorComments ann of
+                                (c, dp) : rem -> (emptyPriorHaddockComment, dp) : (c, DP (2,0)) : rem
+                                _ -> [(emptyPriorHaddockComment, annEntryDelta ann)],
+                            annEntryDelta = DP (1, dpCol)
+                        }
+                 in Map.adjust updateCurrent (mkAnnKey node)
+
+-- | Determine if a list of constructor declarations is missing some haddock comments.
+missingSomeHaddock :: Anns -> [LConDecl GhcPs] -> Bool
+missingSomeHaddock anns = any $ \lcon@(L _ conDecl) -> case conDecl of
+    ConDeclH98 { con_args = RecCon (L _ fields) } ->
+        elem (Just False) $ hasHaddock anns lcon : fmap (hasHaddock anns) fields
+    _                                       -> False -- GADT is not supported yet
+
+-- | Returns 'True' if the end of the first node and the start of the second node are on the same line.
+notSeperatedByLineEnding :: Located a
+                         -> Located a
+                         -> Bool
+notSeperatedByLineEnding (L (RealSrcSpan x _) _) (L (RealSrcSpan y _) _) =
+    srcLocLine (realSrcSpanEnd x) == srcLocLine (realSrcSpanStart y)
+notSeperatedByLineEnding _ _ = False
+
+-- | Empty haddock, suitable for being added to 'annPriorComments'
+emptyPriorHaddockComment :: Comment
+emptyPriorHaddockComment = mkComment "-- |"
+#if MIN_VERSION_ghc(9,0,0)
+    badRealSrcSpan
+#else
+    noSrcSpan
+#endif
+
+-- | Determines the given node has haddock comments attached to it.
+hasHaddock :: Data a => Anns -> Located a -> Maybe Bool
```

Comment:
Should this just return `False` in the `Nothing` case? Or is it useful to be able to distinguish "don't know"?

Reply:
I think it's better to leave the caller to decide what to do when the given `Located a` is missing in the `Anns`. For example, the caller use `False` as the condition to enable the action for adding new haddock comments, but if it's `Nothing`, we should disable this action.